torch/csrc/jit/tensorexpr/cuda_codegen.cpp

```diff
@@ -682,176 +712,139 @@ Stmt* GPUMetaVarRewriter::mutate(const For* v) {
 
     const Var* metaVar = gpu_thread_vars_[gpu_thread_index];
     body = Substitute(Stmt::clone(body), {{v->var(), metaVar}});
+  }
 
-    // extents is the max found, if we don't equal that we may be smaller.
-    // Mask it with the current loops stop.
-    if (!exprEquals(extents[gpu_thread_index], v->stop())) {
-      body = new Cond(
-          new CompareSelect(
-              metaVar,
-              v->stop(),
-              new IntImm(1),
-              new IntImm(0),
-              CompareSelectOperation::kLT),
-          body,
-          nullptr);
-    }
+  // Recurse into body block.
+  body = Stmt::clone(body->accept_mutator(this));
+
+  // pop the internal rank off the stack.
+  if (loop_options.is_gpu_block_index()) {
+    current_block_rank_[loop_options.gpu_block_index()] = old_rank;
+    return body;
+  } else if (loop_options.is_gpu_thread_index()) {
+    current_thread_rank_[loop_options.gpu_thread_index()] = old_rank;
     return body;
   }
 
-  return new For(
-      v->var(), v->start(), v->stop(), Stmt::clone(body), loop_options);
+  return v->cloneWithNewBody(body);
 }
 
 Stmt* GPUMetaVarRewriter::mutate(const Block* v) {
-  bool any_change = false;
+  std::vector<Segment> innerSegments;
+  Segment current;
 
-  std::vector<Stmt*> stmts;
+  auto pushAndReset = [&](bool unmask) {
```

Comment:
Minor: the term "unmask" is a bit hard to understand as a name. From the first impression, I wasn't sure whether it is a verb, or a noun. Maybe "no_mask", or the opposite "need_mask"?